scripts/solve_network.py

```diff
@@ -774,84 +782,72 @@ def add_chp_constraints(n):
 
+    # output ratio between heat and electricity and top_iso_fuel_line for extendable
     if not electric_ext.empty:
-        link_p_nom = get_var(n, "Link", "p_nom")
-
-        # ratio of output heat to electricity set by p_nom_ratio
-        lhs = linexpr(
-            (
-                n.links.loc[electric_ext, "efficiency"]
-                * n.links.loc[electric_ext, "p_nom_ratio"],
-                link_p_nom[electric_ext],
-            ),
-            (-n.links.loc[heat_ext, "efficiency"].values, link_p_nom[heat_ext].values),
-        )
+        p_nom = n.model["Link-p_nom"]
 
-        define_constraints(n, lhs, "=", 0, "chplink", "fix_p_nom_ratio")
-
-        # top_iso_fuel_line for extendable
-        lhs = linexpr(
-            (1, link_p[heat_ext]),
-            (1, link_p[electric_ext].values),
-            (-1, link_p_nom[electric_ext].values),
+        lhs = (
+            p_nom.loc[electric_ext]
+            * (n.links.p_nom_ratio * n.links.efficiency)[electric_ext].values
+            - p_nom.loc[heat_ext] * n.links.efficiency[heat_ext].values
         )
+        n.model.add_constraints(lhs == 0, name="chplink-fix_p_nom_ratio")
 
-        define_constraints(n, lhs, "<=", 0, "chplink", "top_iso_fuel_line_ext")
+        rename = {"Link-ext": "Link"}
+        lhs = (
+            p.loc[:, electric_ext]
+            + p.loc[:, heat_ext]
+            - p_nom.rename(rename).loc[electric_ext]
+        )
+        n.model.add_constraints(lhs <= 0, name="chplink-top_iso_fuel_line_ext")
 
+    # top_iso_fuel_line for fixed
     if not electric_fix.empty:
-        # top_iso_fuel_line for fixed
-        lhs = linexpr((1, link_p[heat_fix]), (1, link_p[electric_fix].values))
-
-        rhs = n.links.loc[electric_fix, "p_nom"].values
-
-        define_constraints(n, lhs, "<=", rhs, "chplink", "top_iso_fuel_line_fix")
+        lhs = p.loc[:, electric_fix] + p.loc[:, heat_fix]
+        rhs = n.links.p_nom[electric_fix]
+        n.model.add_constraints(lhs <= rhs, name="chplink-top_iso_fuel_line_fix")
 
+    # back-pressure
     if not electric.empty:
-        # backpressure
-        lhs = linexpr(
-            (
-                n.links.loc[electric, "c_b"].values * n.links.loc[heat, "efficiency"],
-                link_p[heat],
-            ),
-            (-n.links.loc[electric, "efficiency"].values, link_p[electric].values),
+        lhs = (
+            p.loc[:, heat] * (n.links.efficiency[heat] * n.links.c_b[electric].values)
+            - p.loc[:, electric] * n.links.efficiency[electric]
         )
-
-        define_constraints(n, lhs, "<=", 0, "chplink", "backpressure")
+        n.model.add_constraints(lhs <= rhs, name="chplink-backpressure")
 
 
 def add_co2_sequestration_limit(n, sns):
     co2_stores = n.stores.loc[n.stores.carrier == "co2 stored"].index
 
-    if co2_stores.empty or ("Store", "e") not in n.variables.index:
+    if co2_stores.empty:  # or ("Store", "e") not in n.variables.index:

```

Comment:
For this fragment, would be great to harmonise treatment of `("Store", "e") not in n.variables.index`. Do you have any ideas in this regard?

Reply:
I think the applies to the comment from ealier comment which is `.variables` is no longer available in new pypsa version.